framework/aster-frame/src/trap/irq.rs

```diff
@@ -118,41 +118,62 @@ pub fn disable_local() -> DisabledLocalIrqGuard {
 
 /// A guard for disabled local IRQs.
 pub struct DisabledLocalIrqGuard {
-    was_enabled: bool,
+    /// Whether the local IRQs need to be re-enabled when this gaurd is dropped.
+    /// Usually depends on whether the local IRQs were enabled at the moment the guard was created.
+    need_reenable: bool,
     preempt_guard: DisablePreemptGuard,
 }
 
 impl !Send for DisabledLocalIrqGuard {}
 
+/// Disable local IRQs if they are enabled.
+/// Return whether the local IRQs were enabled before this function call.
+pub fn disable_irq_if_enabled() -> bool {
```

Comment:
I have two suggestions.

First, the pair of `disable_irq_if_enabled` and `enable_irq_if_disabled` are easy to misuse compared to `disable_local`. Please keep them as private functions unless there is a strong evidence that the users of the Framework need them.

Second, `disable_irq_if_enabled` is not a particular good name. Technically, the `disable_local` function also disables IRQs only when the IRQs are not disabled already. So the function name `disable_irq_if_enabled` fails to do a good job on distinguishing it from `disable_local`. If the function shall be kept as a public function, I suggest calling it `disable_local_noguard`. If the function is made private, I suggest calling it `do_disable_local`, whose name emphasizes that the function should only be used internally, with caution.

Reply:
AFIK, this function is not `disable_local_noguard`. Currently `disable_local` actually does two things, disabling preemption _and_ disabling local IRQs. This function wants to disable IRQs without disabling preemption.

In reality, if local IRQs were disabled, there would be no preemption at all. However, I believe the preemption is still disabled because we want to cause further `schedule()` calls to panic, as it is usually not correct to do task switching when preemption or IRQs are disabled.

`preempt()` is special in the way that it disables IRQs, but it _does_ want to do task switching. Since disabling preemption cause panics in `switch_task()`, this PR introduces a function that disables IRQs without disabling preemption.

This is really confusing and it is hard to learn the difference from the function name. So perhaps we need some better API designs here.

Reply:
> AFAIK, this function is not `disable_local_noguard`. Currently `disable_local` actually does two things, disabling preemption _and_ disabling local IRQs. This function wants to disable IRQs without disabling preemption.

Yes, you are right. But that doesn't change the fact the function's name is confusing and should be kept private.

> preempt() is special in the way that it disables IRQs, but it does want to do task switching. Since disabling preemption cause panics in switch_task(), this PR introduces a function that disables IRQs without disabling preemption.

My fundamental concern with this PR is that to prevent the panic caused by the preemption guard in `switch_to_task`, the simplest fix to me is to remove the call to `disable_local` in `preempt`. The kernel preemption (as opposite to user preemption) is not the top of our TODO list for improving the scheduling subsystem.

Reply:
> My fundamental concern with this PR is that to prevent the panic caused by the preemption guard in switch_to_task, the simplest fix to me is to remove the call to disable_local in preempt. The kernel preemption (as opposite to user preemption) is not the top of our TODO list for improving the scheduling subsystem.

Just removing the `disable_local` call and leaving a TODO here is also fine for me. But this PR and the related issue makes me think about the API design around `disable_local()` and `disable_preempt()`. Perhaps making `disable_local()` also disable preemption is not a good idea. In contrast, we can always do the following:
 - Make `disable_local()` _not_ disable preemption.
 - Add a `might_sleep()` function that checks that both IRQs and preemption are _not_ disabled.
   - Otherwise, we'll panic.
 - Remove the `panic!()` check in `switch_to_task()`.
   - Instead, call `might_sleep()` in the beginning of `preempt()` and `schedule()`.

Compared to the alternatives, i.e., adding the third guard that only disables IRQs without disabling preemption, I think the above proposal achieves a clearer API design.

`might_sleep()` can be useful as it can be called in many other places such as `Mutex::lock()`, to ensure performing the context switching is always safe whenever `Mutex::lock()` is called.

Indeed, I've proposed a similar idea in the original issue last year ([this comment](https://github.com/asterinas/asterinas/issues/478#issuecomment-1807197636)). But it's somehow outside the scope of this PR anyway.

Reply:
I am ok with introducing the atomic mode and `might_break` in another PR. But I am not convinced this thing of "Make disable_local() not disable preemption." Let's discuss it further in the other PR.